### PR TITLE
perf(ext/http): optimize for zero or one-packet response streams

### PR DESCRIPTION
**Pull request by mmastrac**
This should improve `deno_reactdom_ssr_flash.jsx`: 39k req/s -> 63k req/s

```
09:29 $ third_party/prebuilt/mac/wrk http://localhost:4500
Running 10s test @ http://localhost:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   251.00us   75.18us   4.43ms   90.92%
    Req/Sec    19.81k   773.52    20.62k    97.52%
  398137 requests in 10.10s, 68.72MB read
Requests/sec:  39429.22
Transfer/sec:      6.81MB

09:52 $ third_party/prebuilt/mac/wrk http://localhost:4500
Running 10s test @ http://localhost:4500
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   155.35us   56.63us   3.84ms   91.88%
    Req/Sec    31.71k     1.48k   33.45k    87.62%
  637327 requests in 10.10s, 98.46MB read
Requests/sec:  63103.75
Transfer/sec:      9.75MB
```

### CODE REVIEW

File: ext/http/00_serve.js [hunk 1]
```diff
@@ -331,11 +331,28 @@ function fastSyncResponseOrStream(req, respBody) {
...
+  if (done1) {
+    core.ops.op_set_promise_complete(req, status);
+    return;
+  }
+  const { value: value2, done: done2 } = await reader.read();
+  if (done2) {
+    core.ops.op_set_response_body_bytes(req, value1);
+    core.ops.op_set_promise_complete(req, status);
+    return;
+  }
```
Comment (lucacasonato) [inline ext/http/00_serve.js:347]: If these reads throw, bad things happen. This should happen inside of the try-catch-finally.

Comment (bartlomieju) [reply ext/http/00_serve.js:347]: Agreed, otherwise LGTM

File: ext/http/00_serve.js [hunk 2]
```diff
@@ -331,25 +331,76 @@ function fastSyncResponseOrStream(req, respBody) {
...
+
+    // The second value cannot block indefinitely, as someone may be waiting on a response
+    // of the first packet that may influence this packet. We set this timeout arbitrarily to 250ms
+    // and we race it.
+    let read2 = reader.read();
+    let timeoutHandle;
+    const timeout = new Promise((r) => {
+      timeoutHandle = setTimeout(r, 250);
+    });
+    const winner = Promise.race([read2, timeout]);
```
Comment (szmarczak) [inline ext/http/00_serve.js:355]: Aren't you missing an await here?

**Comment by ry** [comment]
massive improvement - nice work

File: ext/http/00_serve.js [hunk 3]
```diff
@@ -331,24 +332,92 @@ function fastSyncResponseOrStream(req, respBody) {
...
       const { value, done } = await reader.read();
       if (done) {
+        closed = true;
         break;
       }
       await core.writeAll(responseRid, value);
     }
   } catch (error) {
-    await reader.cancel(error);
+    console.trace(error);
```
Comment (petamoriken) [inline ext/http/00_serve.js:404]: Code for debug is still here

File: ext/http/00_serve.js (hunk 2 above)
Comment (mmastrac) [reply ext/http/00_serve.js]: Newer version of the code changes this race.

File: ext/http/00_serve.js (hunk 1 above)
Comment (mmastrac) [reply ext/http/00_serve.js]: I reworked this pretty massively and added throwing stream tests.

**Review by bartlomieju — APPROVED** [review]
LGTM, a bit involved, but the perf gains are too good to skip.

**Pull request merged**